src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl

```diff
@@ -0,0 +1,541 @@
+    lt,tail = Iterators.peel(terms(g,ordering=ordering));
+    d = dot(u,fmpz.(leading_exponent_vector(lt)))
+
+    initial_terms = [s for s in tail if dot(u,fmpz.(leading_exponent_vector(s)))==d]
+    push!(initial_terms,lt)
+    return sum(initial_terms)
+
+end
+# Applies the function above to each polynomial in a list of polynomials.
+function initial(G::Vector{<:MPolyElem}, ord::MonomialOrdering, u::Vector{fmpz})
+    return initial.(G,Ref(ord),Ref(u))
+end
+
+
+
+# Assumes:
+# - G is a Groebner basis with respect to ord
+# Returns:
+# - a reduced Groebner basis with respect to ord
+function interreduce(G::Vector{<:MPolyElem}, ord::MonomialOrdering)
+
+    # sort G by its leading monomials (result is smallest first)
+    # then reduce G[i] by G[1:i-1] for i=length(G),...,2
+    sort!(G,
+          by=g->leading_monomial(g,ordering=ord),
+          lt=(f,g)->(cmp(ord,f,g)<0))
+    for i in 2:length(G)
+        G[i] = reduce(G[i],G[1:i-1],ordering=ord)
+    end
+    return G
+
+end
+
+
+
+# Assumes:
+# - Hnew is a Groebner basis of the initial ideal w.r.t. ordH
+# - G is a Groebner basis of the original ideal w.r.t. ordG
+# - ordH and ordG are adjacent orderings
+# Returns:
+# - Gnew, a Groebner basis of the original ideal w.r.t. ordH
+function groebner_lift(Hnew::Vector{<:MPolyElem},
+                       ordH::MonomialOrdering,
+                       G::Vector{<:MPolyElem},
+                       ordG::MonomialOrdering)
+
+    # lift to groebner basis
+    Gnew = Hnew - reduce(Hnew,G,ordering=ordG)
+    # interreduce to make reduced
+    Gnew = interreduce(Gnew,ordH)
+    return Gnew
+
+end
+
+
+
+# Assumes:
+# - G is a Groebner basis with respect to ordG
+# - u is an interior facet point of the maximal Groebner cone around ordG
+# - v is an outer normal vector of the facet containing u
+# Returns:
+# - the reduced Groebner basis with respect to the adjacent ordering
+#     in direction v
+function groebner_flip(G::Vector{<:MPolyElem},
+                       ordG::MonomialOrdering,
+                       homogeneityVector::Union{Vector{fmpz},Nothing},
+                       interior_facet_point::Vector{fmpz},
+                       outer_normal_vector::Vector{fmpz})
+
+    R = parent(first(G))
+    n = ngens(R)
+    adjacentOrdering = groebner_flip_adjacent_ordering(R,homogeneityVector,interior_facet_point,outer_normal_vector)
+    H = initial(G,ordG,interior_facet_point)
+    Hnew = collect(groebner_basis(ideal(H),ordering=adjacentOrdering,complete_reduction=true))
+    Gnew = groebner_lift(Hnew,adjacentOrdering,G,ordG)
+    return Gnew,adjacentOrdering
+
+end
+# helper functions for groebner_flip
+function groebner_flip_adjacent_ordering(R::MPolyRing,
+                                         homogeneityVector::Vector{fmpz},
+                                         interior_facet_point::Vector{fmpz},
+                                         outer_normal_vector::Vector{fmpz})
+
+    return weight_ordering(Int.(homogeneityVector),
```

Comment:
Wait, so does that mean all those `Vector{fmpz}` can be safely converted to `Vector{Int}`? So why don't we store them as `Vector{Int}` in the first place?

Reply:
A lot of objects in the tropical code lives in the intersection between polymake and singular.  The former uses `fmpz` and the latter uses `int`.   Since changing polymake to `int` makes less sense to me than teaching Singular to use `fmpz`, I decided to keep everything as `fmpz` and convert to `int` only when necessary.  It's easier to remember than deciding on a case-by-case basis when to use `fmpz` and when to use `int`.